File: frontend/src/lib/triage.js
import { REGIONS, TRIGGERS, DURATIONS } from '../data/regions.js';
import { screenText } from './redflags.js';

/**
 * Deterministic triage scoring.
 *
 * Determinism is the point: a model that "usually" escalates chest pain is not a
 * safety rail. The language model never decides urgency — it only rewrites the
 * output of this function into plain language. Same input, same band, always.
 * The backend runs the same rules (app/services/triage.py); the two are kept in
 * step by src/test/parity.test.js.
 */

export const BANDS = {
  low: { id: 'low', label: 'Low — keep monitoring', min: 0 },
  moderate: { id: 'moderate', label: 'Moderate — see someone soon', min: 30 },
  immediate: { id: 'immediate', label: 'Immediate care needed', min: 65 },
};

export const BAND_RANK = { low: 0, moderate: 1, immediate: 2 };

export function bandFor(score) {
  if (score >= BANDS.immediate.min) return 'immediate';
  if (score >= BANDS.moderate.min) return 'moderate';
  return 'low';
}

const weightOf = (list, id) => list.find((x) => x.id === id)?.weight ?? 0;

/** Why a single red-flag symptom is treated as an emergency on its own. */
const RED_FLAG_REASONS = {
  droop: 'Facial drooping or slurred speech is a stroke red flag. Call emergency services now — minutes change the outcome.',
  worst_headache: 'A sudden, worst-ever headache needs emergency assessment even if it eases off.',
  radiating: 'Chest pain spreading to the arm or jaw is treated as a possible heart attack until proven otherwise.',
  blood: 'Blood in stool or vomit always needs same-day medical assessment.',
  fainting: 'Fainting or collapse needs urgent in-person assessment.',
};

/** Combination rules that escalate beyond the sum of their parts. */
const ESCALATION_RULES = [
  {
    id: 'cardiac_exertional',
    test: ({ region, symptoms, triggers }) =>
      region === 'chest' && symptoms.includes('pressure') && (symptoms.includes('breath') || triggers.includes('exertion')),
    reason: 'Chest pressure together with breathlessness or exertion is treated as an emergency pattern, not a wait-and-see combination.',
  },
  {
    id: 'cardiac_at_rest',
    test: ({ region, symptoms, triggers, severity }) =>
      region === 'chest' && symptoms.includes('pressure') && triggers.includes('rest') && severity >= 6,
    reason: 'Significant chest pressure occurring at rest needs urgent assessment.',
  },
  {
    id: 'acute_abdomen',
    test: ({ region, symptoms, severity }) => region === 'abdomen' && symptoms.includes('rigid') && severity >= 7,
    reason: 'A rigid, severely painful abdomen needs urgent in-person assessment.',
  },
  {
    id: 'systemic_infection',
    test: ({ region, symptoms, severity }) => region === 'general' && symptoms.includes('fever') && severity >= 8,
    reason: 'High fever with severe illness needs same-day assessment.',
  },
  {
    id: 'red_flag_combo',
    test: ({ symptoms }) => symptoms.includes('weight_loss') && symptoms.includes('night_sweats'),
    reason: 'Unintentional weight loss with night sweats should be reviewed by a clinician promptly.',
  },
];

/**
 * @param {{region: string, symptoms?: string[], triggers?: string[], severity?: number, duration?: string, notes?: string}} input
 * @returns {{score: number, band: string, redFlag: boolean, reasons: string[], rules: string[], breakdown: object}}
 */
export function scoreTriage(input) {
  const { region, symptoms = [], triggers = [], severity = 1, duration = 'days', notes = '' } = input;
  const catalogue = REGIONS[region]?.symptoms ?? [];

  let symptomScore = 0;
  let redFlag = false;
  const reasons = [];
  const rules = [];

  for (const s of catalogue) {
    if (!symptoms.includes(s.id)) continue;
    symptomScore += s.weight;
    if (s.redFlag) {
      redFlag = true;
      if (RED_FLAG_REASONS[s.id]) reasons.push(RED_FLAG_REASONS[s.id]);
    }
  }

  const triggerScore = triggers.reduce((sum, id) => sum + weightOf(TRIGGERS, id), 0);
  const durationScore = weightOf(DURATIONS, duration) || 8;
  const severityScore = severity * 3;

  for (const rule of ESCALATION_RULES) {
    if (rule.test({ region, symptoms, triggers, severity })) {
      redFlag = true;
      rules.push(rule.id);
      reasons.push(rule.reason);
    }
  }

  // Free text is screened too: the notes box is where people describe what the
  // tick-boxes could not.
  const text = screenText(notes);
  if (text.triggered) {
    redFlag = true;
    rules.push(...text.categories.map((c) => `text:${c}`));
    reasons.push(...text.reasons);
  }

  const raw = symptomScore + triggerScore + durationScore + severityScore;
  const score = redFlag ? 100 : Math.max(0, Math.min(100, raw));

  return {
    score,
    band: bandFor(score),
    redFlag,
    reasons,
    rules,
    breakdown: { symptomScore, triggerScore, durationScore, severityScore },
  };
}

/**
 * Discussion points, never diagnoses. Wording matters here.
 * Statements about the person's own conditions only appear if they listed them.
 */
export function considerationsFor(input, result) {
  const { region, symptoms = [], triggers = [], conditionIds = [] } = input;
  const has = (id) => symptoms.includes(id);
  const trig = (id) => triggers.includes(id);
  const cond = (id) => conditionIds.includes(id);
  const out = result.reasons.map((text) => ({ text, warn: true }));

  if (region === 'chest') {
    if (has('burning') && (trig('meals') || trig('lying')))
      out.push({
        text: cond('gerd')
          ? 'Burning that follows meals or lying down fits the reflux you have listed. Worth raising, not worth panicking over.'
          : 'Burning that follows meals or lying down is often reflux. Worth raising if it keeps coming back.',
      });
    if (has('palpitations'))
      out.push({ text: 'Note when the palpitations happen and, if you have a monitor, your heart rate at the time. That is what a clinician will ask.' });
    if (has('cough') && trig('night'))
      out.push({ text: 'A night cough can be reflux reaching the throat rather than a chest problem. Mention both to the clinician.' });
  }
  if (region === 'abdomen') {
    if (has('upper_pain') && trig('meals'))
      out.push({ text: cond('gerd') ? 'Upper abdominal pain after eating belongs in the same conversation as your reflux.' : 'Upper abdominal pain after eating is worth mentioning to a clinician if it repeats.' });
    if (has('nausea') && has('bloating') && cond('t2d'))
      out.push({ text: 'Nausea with bloating can follow slow stomach emptying, which is more common with long-standing diabetes. Mention the pairing.' });
  }
  if (region === 'head' && (has('dizzy') || has('blurred')))
    out.push({
      text: cond('t2d') || cond('htn')
        ? 'Dizziness and vision changes track with both blood sugar and blood pressure swings. Check both before your next log.'
        : 'Dizziness and vision changes can follow blood pressure or blood sugar swings. If you have a home monitor, check both.',
    });
  if (region === 'joints') {
    if (has('swelling'))
      out.push({ text: 'A hot, swollen joint is worth seeing within a day or two rather than resting and watching.' });
    if (has('numbness'))
      out.push({ text: cond('t2d') ? 'Numbness in the feet or legs sits on the diabetic review checklist. Flag it at your next appointment.' : 'Numbness or pins and needles that keep returning are worth mentioning to a clinician.' });
  }
  if (region === 'general') {
    if (has('fever'))
      out.push({ text: 'A fever that lasts more than a few days, or comes back after settling, is worth a clinician\'s look.' });
    if (has('weight_loss'))
      out.push({ text: 'Weight loss you did not plan is worth raising even when nothing else hurts.' });
  }

  if (!out.length)
    out.push({ text: 'Nothing in this combination matches an escalation rule. Keep logging so the pattern has something to sit against.' });

  return out;
}

export function nextStepFor(band) {
  if (band === 'immediate')
    return {
      title: 'Get emergency care now',
      body: 'Call your local emergency number or go to the nearest emergency department. Do not drive yourself.',
      tone: 'crimson',
    };
  if (band === 'moderate')
    return {
      title: 'Book a consult in the next 24 to 48 hours',
      body: 'Take this summary with you — the Care team page can copy it. Come sooner if anything sharpens or spreads.',
      tone: 'sand',
    };
  return {
    title: 'Self-care and monitoring look reasonable',
    body: 'Log again tonight. Come back sooner if anything sharpens, spreads, or wakes you up.',
    tone: 'sage',
  };
}

/**
 * Folds the backend's answer into a result the browser already scored.
 * The band can only go up: if the server saw something the local rules did not
 * (a wording the text screen missed, say) the result is escalated; it is never
 * softened. The plain-language summary and citations are always taken.
 */
export function mergeServerResult(local, server) {
  const patch = { summary: server.summary ?? null, citations: server.citations ?? [], source: 'server' };
  if ((BAND_RANK[server.band] ?? 0) > (BAND_RANK[local.band] ?? 0)) {
    return {
      ...patch,
      band: server.band,
      score: server.score,
      redFlag: Boolean(server.red_flag),
      findings: server.findings?.length ? server.findings : local.findings,
      breakdown: server.breakdown
        ? {
            symptomScore: server.breakdown.symptoms ?? 0,
            triggerScore: server.breakdown.triggers ?? 0,
            durationScore: server.breakdown.duration ?? 0,
            severityScore: server.breakdown.severity ?? 0,
          }
        : local.breakdown,
    };
  }
  return patch;
}

/** Routes free text to a page. Deterministic, keyword-first, no model call. */
export function routeQuery(text = '') {
  const q = text.toLowerCase().trim();
  if (!q) return null;
  const map = [
    { to: '/triage', keys: ['pain', 'ache', 'headache', 'chest', 'dizzy', 'nausea', 'symptom', 'knee', 'swollen', 'breath'] },
    { to: '/nutrition', keys: ['eat', 'food', 'meal', 'rice', 'ramen', 'sodium', 'carb', 'sugar', 'toast', 'coffee', 'calorie'] },
    { to: '/vitals', keys: ['heart rate', 'hrv', 'glucose', 'sleep', 'anomaly', 'wearable', 'bpm', 'blood pressure', 'weight'] },
    { to: '/care', keys: ['doctor', 'medicine', 'medication', 'pill', 'appointment', 'interaction', 'clinician'] },
    { to: '/library', keys: ['read', 'article', 'screening', 'checkup', 'prevent'] },
  ];
  for (const row of map) if (row.keys.some((k) => q.includes(k))) return row.to;
  return null;
}

File: backend/app/services/triage.py
"""Deterministic triage scoring plus a grounded, non-diagnostic explanation.

The model never decides the band. `score_triage()` is a pure function: same
input, same output, forever. The reasoning model receives a band that has
already been decided and rewrites it. When the band is `immediate` there is no
model call at all — an emergency is the wrong moment for a network round-trip.
"""

from __future__ import annotations

import time

from synora_final.backend.app.nim import NimError, NimNotConfigured, nim
from synora_final.backend.app.nim.parsing import parse_json
from synora_final.backend.app.nim.prompts import TRIAGE_EXPLAINER
from synora_final.backend.app.rag.retriever import format_context, retrieve
from synora_final.backend.app.safety.redflags import EMERGENCY_TEXT, screen_text

DISCLAIMER = (
    "Triage guidance only. This is not a diagnosis and does not replace a clinician. "
    "In an emergency, call your local emergency number."
)

SYMPTOMS: dict[str, dict[str, dict]] = {
    "head": {
        "headache": {"w": 8, "label": "Headache"},
        "dizzy": {"w": 12, "label": "Dizziness"},
        "blurred": {"w": 16, "label": "Blurred vision"},
        "light": {"w": 8, "label": "Light sensitivity"},
        "droop": {"w": 100, "label": "Facial droop or slurred speech", "red": True},
        "worst_headache": {"w": 100, "label": "Worst headache of my life, sudden", "red": True},
    },
    "chest": {
        "pressure": {"w": 34, "label": "Pressure or tightness"},
        "breath": {"w": 30, "label": "Short of breath"},
        "palpitations": {"w": 18, "label": "Racing or skipping beats"},
        "burning": {"w": 8, "label": "Burning behind the breastbone"},
        "cough": {"w": 8, "label": "Persistent cough"},
        "radiating": {"w": 100, "label": "Pain spreading to arm or jaw", "red": True},
    },
    "abdomen": {
        "nausea": {"w": 10, "label": "Nausea"},
        "bloating": {"w": 6, "label": "Bloating after eating"},
        "upper_pain": {"w": 18, "label": "Pain under the ribs"},
        "reflux": {"w": 8, "label": "Reflux when lying down"},
        "rigid": {"w": 34, "label": "Belly hard and painful"},
        "blood": {"w": 100, "label": "Blood in stool or vomit", "red": True},
    },
    "joints": {
        "knee_pain": {"w": 8, "label": "Joint pain on movement"},
        "stiffness": {"w": 8, "label": "Morning stiffness"},
        "swelling": {"w": 16, "label": "Swollen or warm to touch"},
        "range": {"w": 10, "label": "Reduced range of motion"},
        "numbness": {"w": 18, "label": "Numbness or tingling"},
    },
    "general": {
        "fever": {"w": 16, "label": "Fever"},
        "fatigue": {"w": 8, "label": "Unusual fatigue"},
        "weight_loss": {"w": 22, "label": "Unintentional weight loss"},
        "night_sweats": {"w": 18, "label": "Night sweats"},
        "fainting": {"w": 100, "label": "Fainting or collapse", "red": True},
    },
}

TRIGGERS = {
    "exertion": 14, "meals": 4, "lying": 4, "stress": 5, "cold": 2, "night": 8, "rest": 10,
}
DURATIONS = {"today": 4, "days": 8, "week": 12, "longer": 16}

# Why a single red-flag symptom is an emergency by itself. Mirrors the wording in
# frontend/src/lib/triage.js so the person reads the same reason either way.
RED_FLAG_REASONS = {
    "droop": "Facial drooping or slurred speech is a stroke red flag. Call emergency services now — minutes change the outcome.",
    "worst_headache": "A sudden, worst-ever headache needs emergency assessment even if it eases off.",
    "radiating": "Chest pain spreading to the arm or jaw is treated as a possible heart attack until proven otherwise.",
    "blood": "Blood in stool or vomit always needs same-day medical assessment.",
    "fainting": "Fainting or collapse needs urgent in-person assessment.",
}

ESCALATION_RULES = [
    (
        "cardiac_exertional",
        lambda r, s, t, sev: r == "chest" and "pressure" in s and ("breath" in s or "exertion" in t),
        "Chest pressure together with breathlessness or exertion is treated as an emergency pattern, "
        "not a wait-and-see combination.",
    ),
    (
        "cardiac_at_rest",
        lambda r, s, t, sev: r == "chest" and "pressure" in s and "rest" in t and sev >= 6,
        "Significant chest pressure occurring at rest needs urgent assessment.",
    ),
    (
        "acute_abdomen",
        lambda r, s, t, sev: r == "abdomen" and "rigid" in s and sev >= 7,
        "A rigid, severely painful abdomen needs urgent in-person assessment.",
    ),
    (
        "systemic_infection",
        lambda r, s, t, sev: r == "general" and "fever" in s and sev >= 8,
        "High fever with severe illness needs same-day assessment.",
    ),
    (
        "red_flag_combo",
        lambda r, s, t, sev: {"weight_loss", "night_sweats"}.issubset(set(s)),
        "Unintentional weight loss with night sweats should be reviewed by a clinician promptly.",
    ),
]

BAND_LABELS = {
    "low": "Low — keep monitoring",
    "moderate": "Moderate — see someone soon",
    "immediate": "Immediate care needed",
}


def band_for(score: int) -> str:
    if score >= 65:
        return "immediate"
    if score >= 30:
        return "moderate"
    return "low"


def score_triage(payload: dict) -> dict:
    """Pure. No I/O, no clock, no randomness — so it is testable and repeatable."""
    region = payload.get("region", "general")
    symptoms = list(payload.get("symptoms") or [])
    triggers = list(payload.get("triggers") or [])
    severity = int(payload.get("severity") or 1)
    duration = payload.get("duration", "days")

    catalogue = SYMPTOMS.get(region, {})
    symptom_score = 0
    red_flag = False
    fired: list[str] = []
    reasons: list[str] = []
    for sid in symptoms:
        entry = catalogue.get(sid)
        if not entry:
            continue
        symptom_score += int(entry["w"])
        if entry.get("red"):
            red_flag = True
            if sid in RED_FLAG_REASONS:
                reasons.append(RED_FLAG_REASONS[sid])

    trigger_score = sum(TRIGGERS.get(t, 0) for t in triggers)
    duration_score = DURATIONS.get(duration, 8)
    severity_score = severity * 3

    for name, test, reason in ESCALATION_RULES:
        if test(region, symptoms, triggers, severity):
            red_flag = True
            fired.append(name)
            reasons.append(reason)

    raw = symptom_score + trigger_score + duration_score + severity_score
    score = 100 if red_flag else max(0, min(100, raw))

    return {
        "score": score,
        "band": band_for(score),
        "red_flag": red_flag,
        "escalation_rules": fired,
        "reasons": reasons,
        "breakdown": {
            "symptoms": symptom_score,
            "triggers": trigger_score,
            "duration": duration_score,
            "severity": severity_score,
        },
    }


def considerations(payload: dict, result: dict) -> list[dict]:
    """Discussion points, never diagnoses. Wording matters more than coverage."""
    region = payload.get("region")
    symptoms = set(payload.get("symptoms") or [])
    triggers = set(payload.get("triggers") or [])
    conditions = set(payload.get("condition_ids") or [])
    out = [{"text": r, "warn": True} for r in result["reasons"]]

    if region == "chest":
        if "burning" in symptoms and (triggers & {"meals", "lying"}):
            text = (
                "Burning that follows meals or lying down fits the reflux you have listed. "
                "Worth raising, not worth panicking over."
                if "gerd" in conditions
                else "Burning that follows meals or lying down is often reflux. Worth raising if it keeps coming back."
            )
            out.append({"text": text, "warn": False})
        if "palpitations" in symptoms:
            out.append({"text": "Note when the palpitations happen and, if you have a monitor, your heart rate "
                                "at the time. That is what a clinician will ask.", "warn": False})
    if region == "abdomen":
        if "nausea" in symptoms and "bloating" in symptoms and "t2d" in conditions:
            out.append({"text": "Nausea with bloating can follow slow stomach emptying, which is more common "
                                "with long-standing diabetes. Mention the pairing.", "warn": False})
    if region == "head" and (symptoms & {"dizzy", "blurred"}):
        out.append({"text": "Dizziness and vision changes track with both blood sugar and blood pressure "
                            "swings. Check both before your next log.", "warn": False})
    if region == "joints" and "numbness" in symptoms and "t2d" in conditions:
        out.append({"text": "Numbness in the feet or legs sits on the diabetic review checklist. Flag it at "
                            "your next appointment.", "warn": False})

    if not out:
        out.append({"text": "Nothing in this combination matches an escalation rule. Keep logging so the "
                            "pattern has something to sit against.", "warn": False})
    return out


def next_step(band: str) -> dict[str, str]:
    if band == "immediate":
        return {
            "title": "Get emergency care now",
            "body": "Call your local emergency number or go to the nearest emergency department. "
                    "Do not drive yourself.",
            "tone": "crimson",
        }
    if band == "moderate":
        return {
            "title": "Book a consult in the next 24 to 48 hours",
            "body": "Bring this summary with you. Come sooner if anything sharpens or spreads.",
            "tone": "sand",
        }
    return {
        "title": "Self-care and monitoring look reasonable",
        "body": "Log again later today. Come back sooner if anything sharpens, spreads, or wakes you up.",
        "tone": "sage",
    }


def _retrieval_query(payload: dict, result: dict) -> str:
    labels = [
        SYMPTOMS.get(payload["region"], {}).get(s, {}).get("label", s)
        for s in payload.get("symptoms") or []
    ]
    return f"{payload['region']} {' '.join(labels)} {' '.join(payload.get('triggers') or [])}".strip()


async def run_triage(payload: dict) -> dict:
    started = time.perf_counter()

    # Free-text red flags are checked before scoring — someone can describe an
    # emergency in the notes box while ticking mild boxes above it.
    note_flags = screen_text(payload.get("notes", ""))
    result = score_triage(payload)
    if note_flags.triggered:
        result["red_flag"] = True
        result["score"] = 100
        result["band"] = "immediate"
        result["escalation_rules"] = result["escalation_rules"] + [f"text:{c}" for c in note_flags.categories]
        result["reasons"] = result["reasons"] + note_flags.reasons

    findings = considerations(payload, result)
    response = {
        "band": result["band"],
        "score": result["score"],
        "red_flag": result["red_flag"],
        "escalation_rules": result["escalation_rules"],
        "findings": findings,
        "next_step": next_step(result["band"]),
        "breakdown": result["breakdown"],
        "citations": [],
        "summary": None,
        "disclaimer": DISCLAIMER,
        "decided_by": "rules",
    }

    if result["band"] == "immediate":
        response["summary"] = EMERGENCY_TEXT
        response["latency_ms"] = round((time.perf_counter() - started) * 1000, 1)
        return response

    if not payload.get("explain", True):
        response["latency_ms"] = round((time.perf_counter() - started) * 1000, 1)
        return response

    hits, _retriever, _ = await retrieve(_retrieval_query(payload, result))
    response["citations"] = [
        {"id": h.id, "title": h.title, "source": h.source, "score": h.score} for h in hits
    ]

    notes = payload.get("notes", "")

    user_block = "\n".join(
        [
            f"Urgency band: {result['band']} (score {result['score']}/100).",
            f"Region: {payload['region']}.",
            f"Symptoms: {', '.join(payload.get('symptoms') or []) or 'none selected'}.",
            f"Triggers: {', '.join(payload.get('triggers') or []) or 'none'}.",
            f"Severity at worst: {payload.get('severity')}/10. Duration: {payload.get('duration')}.",
            f"Known conditions: {', '.join(payload.get('condition_ids') or []) or 'none on file'}.",
            f'Their own words: "{notes}"' if notes else "",
            "",
            "CONTEXT:",
            format_context(hits),
        ]
    )

    try:
        raw, _usage = await nim.complete(
            messages=[
                {"role": "system", "content": TRIAGE_EXPLAINER},
                {"role": "user", "content": user_block},
            ],
            temperature=0.2,
            max_tokens=nim.s.nim_max_tokens_triage,
            json_only=True,
        )
        parsed = parse_json(raw, {})
        summary = str(parsed.get("summary") or "").strip()
        if summary:
            response["summary"] = summary
    except (NimNotConfigured, NimError):
        response["summary"] = None  # rules-only result is still a complete answer

    response["latency_ms"] = round((time.perf_counter() - started) * 1000, 1)
    return response
